python/xir/program.py

```diff
@@ -563,6 +578,65 @@ def serialize(self, minimize: bool = False) -> str:
             return strip(res_script)
         return res_script
 
+    @staticmethod
+    def merge(*programs) -> "XIRProgram":
+        """Merges one or more XIR programs into a new XIR program.
+
+        The merged XIR program is formed by concatenating the given XIR programs
+        in the order they are passed to the function. Warnings may be issued for
+        duplicate declarations, gates, includes, operators, and options.
+
+        Args:
+            programs (XIRProgram): XIR programs to merge
+
+        Returns:
+            XIRProgram: the merged XIR program
+
+        Raises:
+            ValueError: if no XIR programs are provided or if at least two XIR
+                programs have different versions or float settings
+        """
+        if len(programs) == 0:
+            raise ValueError("Merging requires at least one XIR program.")
+
+        version = programs[0].version
+        if any(program.version != version for program in programs):
+            raise ValueError("XIR programs with different versions cannot be merged.")
+
+        use_floats = programs[0].use_floats
+        if any(program.use_floats != use_floats for program in programs):
+            raise ValueError("XIR programs with different float settings cannot be merged.")
+
+        result = XIRProgram(version=version, use_floats=use_floats)
+
+        for program in programs:
```

Comment:
It's worth noting that the resulting `XIRProgram` will not be a deep copy of the input programs due to the way the mutator methods are implemented. For example, when adding a statement using `add_statement` the `Statement` object is just referenced, so changing the (private) parameters or wires of the statement changes it in the resulting program as well). Although the `Statement` should be immutable, so it's not necessarily an issue, it's still something that I'm always a bit wary about.

Reply:
Yes, good point! I've created a reminder [ch8135] to update the `XIRProgram` class as this change will be non-trivial: `__eq__` will most likely need to be implemented for all other classes in program.py (for testing purposes).